src/FlowtideDotNet.Core/Compute/Internal/FlowtideExpressionVisitor.cs

```diff
@@ -47,23 +48,27 @@ public FlowtideExpressionVisitor(IFunctionsRegister functionsRegister, System.Ty
 
         public static System.Linq.Expressions.Expression VisitInnerReferenceSegment(ReferenceSegment referenceSegment, System.Linq.Expressions.Expression expression)
         {
-            System.Linq.Expressions.Expression? expr = null;
-            if (referenceSegment is StructReferenceSegment structReferenceSegment)
-            {
-                var method = expression.Type.GetMethod("GetVectorValue");
-                Debug.Assert(method != null);
-                expr = System.Linq.Expressions.Expression.Call(expression, method, System.Linq.Expressions.Expression.Constant(structReferenceSegment.Field));
-            }
-            if (referenceSegment is MapKeyReferenceSegment mapKeyReferenceSegment)
-            {
-                var method = expression.Type.GetMethod("GetMapValue");
-                Debug.Assert(method != null);
-                expr = System.Linq.Expressions.Expression.Call(expression, method, System.Linq.Expressions.Expression.Constant(mapKeyReferenceSegment.Key));
-            }
-            if (expr == null)
+            // Since ref structs cant implement interfaces, check that methods exist in both flx value types
+            Debug.Assert(nameof(FlxValue.GetVectorValue) != null);
+            Debug.Assert(nameof(FlxValueRef.GetVectorValue) != null);
+            Debug.Assert(nameof(FlxValue.GetMapValue) != null);
+            Debug.Assert(nameof(FlxValueRef.GetMapValue) != null);
+
+            var expr = referenceSegment switch
             {
-                throw new NotImplementedException();
-            }
+                StructReferenceSegment { Field: >= 0 } structReferenceSegment => 
+                    System.Linq.Expressions.Expression.Call(
+                        expression, 
+                        expression.Type.GetMethod(nameof(FlxValue.GetVectorValue)) ?? throw new MissingMethodException(nameof(FlxValue)), 
+                        System.Linq.Expressions.Expression.Constant(structReferenceSegment.Field)),
+                MapKeyReferenceSegment { Key: not null } mapKeyReferenceSegment => 
+                    System.Linq.Expressions.Expression.Call(
+                        expression, 
+                        expression.Type.GetMethod(nameof(FlxValue.GetMapValue)) ?? throw new MissingMethodException(nameof(FlxValue)), 
+                        System.Linq.Expressions.Expression.Constant(mapKeyReferenceSegment.Key)),
+                _ => throw new NotImplementedException(),
```

Comment:
```suggestion
                _ => throw new NotImplementedException("The ${nameof(referenceSegment)} must be a {nameof(StructReferenceSegment)} with a positive {nameof(StructReferenceSegment.Field}" or a ${nameof(MapKeyReferenceSegment)} with a non-null ${nameof(MapKeyReferenceSegment.Key)}",
```

What do you think about that? Some way for the caller to get a bit more information why they got a `NotImplementedException`.